src/codemagic/universal_apk_generator.py

```diff
@@ -0,0 +1,166 @@
+        argparse_kwargs={'required': False, 'default': models.Bundletool.DEFAULT_PATH},
+    )
+
+    KEYSTORE_PATH = cli.ArgumentProperties(
+        flags=('--ks',),
+        key='keystore_path',
+        type=pathlib.Path,
+        description='path to the keystore to sign the apk files with',
+        argparse_kwargs={'required': False, 'default': None},
+    )
+
+    KEYSTORE_PASSWORD = cli.ArgumentProperties(
+        flags=('--ks-pass',),
+        key='keystore_password',
+        description='keystore password',
+        argparse_kwargs={'required': False, 'default': None},
+    )
+
+    KEY_ALIAS = cli.ArgumentProperties(
+        flags=('--ks-key-alias',),
+        key='key_lias',
+        description='keystore key alias',
+        argparse_kwargs={'required': False, 'default': None},
+    )
+
+    KEY_PASSWORD = cli.ArgumentProperties(
+        flags=('--key-pass',),
+        key='key_password',
+        description='keystore key password',
+        argparse_kwargs={'required': False, 'default': None},
+    )
+
+
+SigningInfo = collections.namedtuple('SigningInfo', 'store_path store_pass key_alias key_pass')
+
+
+@cli.common_arguments(
+    UniversalApkGeneratorArgument.PATTERN,
+    UniversalApkGeneratorArgument.BUNDLETOOL_PATH,
+    UniversalApkGeneratorArgument.KEYSTORE_PATH,
+    UniversalApkGeneratorArgument.KEYSTORE_PASSWORD,
+    UniversalApkGeneratorArgument.KEY_ALIAS,
+    UniversalApkGeneratorArgument.KEY_PASSWORD
+)
+class UniversalApkGenerator(cli.CliApp):
+    """
+    Generate universal APK files from Android App Bundles
+    """
+
+    def __init__(self, bundletool_path: pathlib.Path, pattern: pathlib.Path, signing_info: Optional[SigningInfo] = None):
+        super().__init__()
+        self.bundletool_path = bundletool_path
+        self.pattern = pattern
+        self.signing_info = signing_info
+
+    @classmethod
+    def from_cli_args(cls, cli_args: argparse.Namespace):
+        keystore_arguments = (
+            UniversalApkGeneratorArgument.KEYSTORE_PATH,
+            UniversalApkGeneratorArgument.KEYSTORE_PASSWORD,
+            UniversalApkGeneratorArgument.KEY_ALIAS,
+            UniversalApkGeneratorArgument.KEY_PASSWORD,
+        )
+        signing_info_args = [getattr(cli_args, a.value.key) for a in keystore_arguments]
+        if any(signing_info_args) and not all(signing_info_args):
+            raise UniversalApkGeneratorError(
+                'either all signing info arguments should be specified, or none of them should')
+
+        pattern = cli_args.pattern.expanduser()
+        if pattern.is_absolute():
+            pattern = pattern.relative_to(pattern.anchor)
+
+        return UniversalApkGenerator(
+            bundletool_path=cli_args.bundletool_path,
+            pattern=pattern,
+            signing_info=SigningInfo(*signing_info_args) if all(signing_info_args) else None,
+        )
+
+    @cli.action('generate')
+    def generate(self) -> NoReturn:
```

Comment:
`NoReturn` is not correct. To make the module useful as Python api, it could even return `apk_paths: List[Path]`.